src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,290 @@
+            {
+                // make sure node hasn't been removed by a different thread
+                if (_map.TryGetValue(key, out var cacheItem))
+                    _eventQueue.Add(() => _doubleLinkedList.Remove(cacheItem));
+
+                // add a new item regardless of whether the old item was removed or not
+                _eventQueue.Add(() => _doubleLinkedList.AddFirst(new LRUCacheItem<TKey, TValue>(key, value, expirationTime)));
+
+                return true;
+            }
+            else
+            {
+                // if cache is full, then remove the least recently used node
+                if (_map.Count >= _capacity)
+                {                
+                    _eventQueue.Add(() =>
+                    {
+                        var lru = _doubleLinkedList.Last;
+                        _map.TryRemove(lru.Value.Key, out _);
+                        _doubleLinkedList.Remove(lru);
+                    });
+                }
+
+                // add a new node
+                var node = new LRUCacheItem<TKey, TValue>(key, value, expirationTime);
+                _eventQueue.Add(() => _doubleLinkedList.AddFirst(node));
+                _map[key] = node;
+
+                return true;
+            }
+        }
+
+        /// Each time a node gets accessed, it gets moved to the beginning (head) of the list.
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (!_map.ContainsKey(key))
+            {
+                value = default;
+                return false;
+            }
+
+            LRUCacheItem<TKey, TValue> cacheItem;
+            // make sure node hasn't been removed by a different thread
+            if (_map.TryGetValue(key, out cacheItem))
+                _eventQueue.Add(() =>
+                {
+                    _doubleLinkedList.Remove(cacheItem);
+                    _doubleLinkedList.AddFirst(cacheItem);
+                });
+
+            value = cacheItem != null ? cacheItem.Value : default;
+            return cacheItem != null;
+        }
+
+        /// Removes a particular key from the cache.
+        public bool TryRemove(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (!_map.ContainsKey(key))
+            {
+                value = default;
+                return false;
+            }
+
+            // check to make sure node wasn't removed by a different thread
+            if (!_map.TryGetValue(key, out var cacheItem))
+            {
+                value = default;
+                return false;
+            }
+
+            _eventQueue.Add(() => _doubleLinkedList.Remove(cacheItem));
+            _map.TryRemove(key, out _);
+            value = cacheItem.Value;
+
+            return true;
```

Comment:
should this be the return value of TryRemove?

Reply:
Yes, that makes sense. Fixed here https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1575/commits/45c168bd46aa16f03a5e5f83c8b9b2b8cef9b070.